node/src/components/block_proposer.rs

```diff
@@ -378,42 +378,38 @@ impl BlockProposerReady {
         past_deploys: HashSet<DeployHash>,
         random_bit: bool,
     ) -> ProtoBlock {
-        let transfers = self
-            .sets
-            .pending
-            .iter()
-            .filter(|(_, deploy_type)| matches!(deploy_type, DeployType::Transfer(_)))
-            .take(deploy_config.block_max_transfer_count as usize);
-        let deploys = self
-            .sets
-            .pending
-            .iter()
-            .filter(|(_, deploy_type)| matches!(deploy_type, DeployType::Wasm(_)))
-            .take(deploy_config.block_max_deploy_count as usize);
-
-        let (deploys, transfers) = deploys.chain(transfers).fold(
-            (Vec::new(), Vec::new()),
-            |(mut deploys, mut transfers), (hash, deploy_type)| {
-                if self.is_deploy_valid(
-                    deploy_type.header(),
-                    block_timestamp,
-                    &deploy_config,
-                    &past_deploys,
-                ) && !past_deploys.contains(hash)
-                    && !self.sets.finalized_deploys.contains_key(hash)
-                {
-                    // all deploys in pending that aren't in finalized blocks or
-                    // proposed blocks from the set `past_blocks`
-                    match deploy_type {
-                        DeployType::Transfer(_) => transfers.push(*hash),
-                        DeployType::Wasm(_) => deploys.push(*hash),
-                    }
+        let max_transfers = deploy_config.block_max_transfer_count as usize;
```

Comment:
I still feel something like this would be simpler and easier to read:
```rust
let is_valid = |(hash, deploy_type): &(DeployHash, DeployType)| {
    self.is_deploy_valid(…) && !past_deploys.contains(hash) && !self.sets.finalized_deploys.contains_key(hash)
};
let transfers = self.sets.pending.iter().filter(|(_, deploy_type)| deploy_type.is_transfer())
    .filter(is_deploy_valid)
    .take(deploy_config.block_max_transfer_count as usize)
    .collect();
let wasm_deploys = self.sets.pending.iter().filter(|(_, deploy_type)| deploy_type.is_wasm())
    .filter(is_deploy_valid)
    .take(deploy_config.block_max_deploy_count as usize)
    .collect();
```

Reply:
An issue with this implementation is that we traverse `sets.pending` twice. In Dan's version we do it once.

Reply:
Yes, that's why I suggested earlier to split up the whole collection.
Anyway, I'm not against the current version either, if you prefer that; it just seems a bit convoluted to me.

Reply:
Yes, if we split the sets into separate ones for `transfers` and `deploys` then you version would be both more readable and equally efficient.